### PR TITLE
Implement P2408R5: Ranges Iterators As Inputs To Non-Ranges Algorithms

### LINKED ISSUE #2925: P2408R5 Ranges Iterators As Inputs To Non-Ranges Algorithms
[P2408R5](https://wg21.link/P2408R5) Ranges Iterators As Inputs To Non-Ranges Algorithms

Feature-test macro (expected):
```cpp
#define __cpp_lib_algorithm_iterator_requirements 202207L
```

### PR DESCRIPTION
**Pull request by strega-nil-ms**
Fixes #2925 

Changes from paper:

* parallel algorithms: require `forward_iterator<FwdIt>` when in concepts mode for read-only iterators
* other algorithms: if there is a `static_assert(_Is_blah_iter_v)`, turn it into a `static_assert(_Is_ranges_blah_iter_v)`, when the iterator is non-mutable
* `sample`: check for `_Is_ranges_fwd_iter_v` 
* `unique_copy`: check for `_Is_ranges_fwd_iter_v`

Other changes, not explicitly specified in the paper:

* many algorithms: if there is a check for different `_Is_blah_iter_v`, and the iterator in question is non-mutable, check `_Is_ranges_blah_iter_v`
  * for example, check for `_Is_ranges_random_iter_v` in `advance`
* `prev`: check for `_Is_ranges_bidi_iter_v`
* `_Use_atomic_iterators` -> check for `_Is_ranges_random_iter_v` - this has ABI impact

Drive by:

* fixes `_Can_reread_dest` to correctly use `_Is_fwd_iter_v`, not checking for `forward_iterator`, as it is a mutable iterator
* add a missed check in `rotate_copy(ExPo)`

Renamings:
* `_REQUIRE_PARALLEL_ITERATOR` becomes two macros:
   + `_REQUIRE_PARALLEL_ITERATOR` checks for `forward_iterator` now, and is used for read-only iterators
   + `_REQUIRE_CPP17_MUTABLE_ITERATOR` checks for `_Is_fwd_iter_v` (like the old `_REQUIRE_PARALLEL_ITERATOR`), but has a new message
* `_Is_blah_iter_v` -> `_Is_cpp17_blah_iter_v` - makes it more obvious that we're checking for classic iterators
* ABI impact from `_Use_atomic_iterators`
   + `_Static_partitioned_find2` -> `_Static_partitioned_find3`
   + `_Static_partitioned_find_end_forward` -> `_Static_partitioned_find_end_forward2`
   + `_Static_partitioned_find_end_backward2` -> `_Static_partitioned_find_end_backward3`
   + `_Static_partitioned_adjacent_find2` -> `_Static_partitioned_adjacent_find3`
   + `_Static_partitioned_mismatch2` -> `_Static_partitioned_mismatch3`
   + `_Static_partitioned_search2` -> `_Static_partitioned_search3`
   + `_Static_partitioned_search_n2` -> `_Static_partitioned_search_n3`
   + `_Static_partitioned_is_sorted_until` -> `_Static_partitioned_is_sorted_until2`
   + `_Static_partitioned_is_heap_until` -> `_Static_partitioned_is_heap_until2`

### CODE REVIEW

File: stl/inc/xutility [hunk 1]
```diff
@@ -848,9 +848,25 @@ _INLINE_VAR constexpr bool _Is_input_iter_v = is_convertible_v<_Iter_cat_t<_Iter
...
+template <class _Iter>
+_INLINE_VAR constexpr bool _Is_ranges_bidi_iter_v =
+#if defined(__cpp_lib_concepts)
+    bidirectional_iterator<_Iter>;
+#else
+    is_convertible_v<_Iter_cat_t<_Iter>, bidirectional_iterator_tag>;
+#endif
+
 template <class _Iter>
 _INLINE_VAR constexpr bool _Is_random_iter_v = is_convertible_v<_Iter_cat_t<_Iter>, random_access_iterator_tag>;
```
Comment (frederick-vs-ja) [inline stl/inc/xutility:871]: Are random access iterators already handled?

**Review by frederick-vs-ja — COMMENTED** [review]
I'm afraid of that reverting improvements introduced by #1794 ...

File: stl/inc/xutility [hunk 2]
```diff
@@ -848,9 +848,25 @@ _INLINE_VAR constexpr bool _Is_input_iter_v = is_convertible_v<_Iter_cat_t<_Iter
 template <class _Iter>
 _INLINE_VAR constexpr bool _Is_fwd_iter_v = is_convertible_v<_Iter_cat_t<_Iter>, forward_iterator_tag>;
 
+template <class _Iter>
+_INLINE_VAR constexpr bool _Is_ranges_fwd_iter_v =
+#if defined(__cpp_lib_concepts)
+    forward_iterator<_Iter>;
+#else
+    is_convertible_v<_Iter_cat_t<_Iter>, forward_iterator_tag>;
+#endif
```
[the remark this answers is not in the document]
Comment (strega-nil-ms) [reply stl/inc/xutility:857]: This is reasonable; the standard does not say we can't allow more types, and therefore not breaking old types is fine.

File: stl/inc/xutility (hunk 1 above)
Comment (strega-nil-ms) [reply stl/inc/xutility:871]: There were no cases where `_Is_ranges_random_iter_v` would have been useful.

**Comment by strega-nil-ms** [comment]
@frederick-vs-ja For this case, we _should_ only be updating `static_assert`s and the specific algorithms specified in the paper, and therefore should not have made any changes that break performance.

File: stl/inc/xutility (hunk 1 above)
Comment (frederick-vs-ja) [reply stl/inc/xutility:871]: Some algorithms use `if constexpr` to detect random access iterators and have lower time complexity whenever possible. 
Perhaps we can also do this for `random_access_iterator`, but this may be out of the scope of P2408R5.

Comment (strega-nil-ms) [reply stl/inc/xutility:871]: right, exactly; I don't wish to make any changes not required by the paper.

File: stl/inc/xutility [hunk 3]
```diff
@@ -1110,7 +1127,7 @@ constexpr void _Debug_order_unchecked(_InIt _First, _Sentinel _Last, _Pr&& _Pred
 template <class _OtherIt, class _InIt, class _Pr>
 constexpr void _Debug_order_set_unchecked(_InIt _First, _InIt _Last, _Pr&& _Pred) {
     // test if range is ordered by predicate
-    if constexpr (is_same_v<_Iter_value_t<_OtherIt>, _Iter_value_t<_InIt>> && _Is_fwd_iter_v<_InIt>) {
+    if constexpr (is_same_v<_Iter_value_t<_OtherIt>, _Iter_value_t<_InIt>>) {
```
Comment (strega-nil-ms) [inline stl/inc/xutility:1130]: Drive by; I noticed that `_Debug_order_unchecked` already checks for `_Is_fwd_iter_v`

File: stl/inc/execution [hunk 4]
```diff
@@ -1576,7 +1576,7 @@ template <class _ExPo, class _FwdIt1, class _FwdIt2, class _Pr, _Enable_if_execu
 _NODISCARD _FwdIt1 find_first_of(_ExPo&& _Exec, const _FwdIt1 _First1, _FwdIt1 _Last1, const _FwdIt2 _First2,
     const _FwdIt2 _Last2, _Pr _Pred) noexcept /* terminates */ {
     // look for one of [_First2, _Last2) that matches element
-    _REQUIRE_PARALLEL_ITERATOR(_FwdIt1);
+    _REQUIRE_CONST_PARALLEL_ITERATOR(_FwdIt1);
```
Comment (StephanTLavavej) [inline stl/inc/execution:1579]: Partially pre-existing: Did we simply forget to check `_FwdIt2`?

File: stl/inc/execution [hunk 5]
```diff
@@ -4388,8 +4388,8 @@ template <class _ExPo, class _FwdIt1, class _FwdIt2, class _Ty, class _BinOp,
 _FwdIt2 exclusive_scan(_ExPo&&, const _FwdIt1 _First, const _FwdIt1 _Last, _FwdIt2 _Dest, _Ty _Val,
     _BinOp _Reduce_op) noexcept /* terminates */ {
     // set each value in [_Dest, _Dest + (_Last - _First)) to the associative reduction of predecessors and _Val
-    _REQUIRE_PARALLEL_ITERATOR(_FwdIt1);
-    _REQUIRE_PARALLEL_ITERATOR(_FwdIt2);
+    _REQUIRE_CONST_PARALLEL_ITERATOR(_FwdIt1);
+    _REQUIRE_CONST_PARALLEL_ITERATOR(_FwdIt2);
```
Comment (StephanTLavavej) [inline stl/inc/execution:4392]: `exclusive_scan()` writes to `_FwdIt2 _Dest`.

File: stl/inc/algorithm [hunk 6]
```diff
@@ -4096,14 +4096,14 @@ namespace ranges {
...
-template <class _InIt, class _OutIt>
-_INLINE_VAR constexpr bool _Can_reread_dest = _Is_fwd_iter_v<_OutIt> //
+#if defined(__cpp_lib_concepts)
+concept _Can_reread_dest
+#else
+_INLINE_VAR constexpr bool _Can_reread_dest
+#endif
+    = _Is_fwd_iter_v<_OutIt> //
     && is_same_v<_Iter_value_t<_InIt>, _Iter_value_t<_OutIt>>;
-#endif // __cpp_lib_concepts
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:4106]: Can you explain what the drive-by fix is? I see `forward_iterator` being replaced with the "stronger" (?) `_Is_fwd_iter_v` for concepts mode.

File: stl/inc/algorithm [hunk 7]
```diff
@@ -4408,7 +4408,7 @@ template <class _ExPo, class _BidIt, class _FwdIt, _Enable_if_execution_policy_t
 _FwdIt reverse_copy(_ExPo&&, _BidIt _First, _BidIt _Last, _FwdIt _Dest) noexcept /* terminates */ {
     // copy reversing elements in [_First, _Last)
     // not parallelized as benchmarks show it isn't worth it
-    _REQUIRE_PARALLEL_ITERATOR(_FwdIt);
+    _REQUIRE_MUTABLE_PARALLEL_ITERATOR(_FwdIt);
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:4411]: We previously didn't check `_BidIt`, presumably because input iterators couldn't be silently passed to something expecting bidi operations.

Now, should we explicitly check `_REQUIRE_CONST_PARALLEL_ITERATOR(_BidIt)`?

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/yvals_core.h [hunk 8]
```diff
@@ -1349,6 +1349,14 @@
 #endif // __cpp_impl_coroutine
 
 #if _HAS_CXX20
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/yvals_core.h:1351]: After reading through this, I think it could be reasonable to implement "downlevel", just want to make sure this is intentional.

[the remark this answers is not in the document]
Comment (frederick-vs-ja) [reply stl/inc/yvals_core.h:1351]: It seems that in some rare cases we should not check concepts (see #2358) unless required by the standard. But I think it's OK to back port the changes to C++20 mode. We can defer the reconsideration until someone complains.

File: stl/inc/xutility [hunk 9]
```diff
@@ -848,9 +848,23 @@ _INLINE_VAR constexpr bool _Is_input_iter_v = is_convertible_v<_Iter_cat_t<_Iter
 template <class _Iter>
 _INLINE_VAR constexpr bool _Is_fwd_iter_v = is_convertible_v<_Iter_cat_t<_Iter>, forward_iterator_tag>;
 
+template <class _Iter>
+_INLINE_VAR constexpr bool _Is_ranges_fwd_iter_v =
+#if defined(__cpp_lib_concepts)
+    forward_iterator<_Iter> ||
+#endif
+    is_convertible_v<_Iter_cat_t<_Iter>, forward_iterator_tag>;
```
Comment (miscco) [inline stl/inc/xutility:856]: I believe it would be easier to read as 
```suggestion
    _Is_fwd_iter_v<_Iter>;
```

File: stl/inc/xutility [hunk 10]
```diff
@@ -848,9 +848,23 @@ _INLINE_VAR constexpr bool _Is_input_iter_v = is_convertible_v<_Iter_cat_t<_Iter
...
+
 template <class _Iter>
 _INLINE_VAR constexpr bool _Is_bidi_iter_v = is_convertible_v<_Iter_cat_t<_Iter>, bidirectional_iterator_tag>;
 
+template <class _Iter>
+_INLINE_VAR constexpr bool _Is_ranges_bidi_iter_v =
+#if defined(__cpp_lib_concepts)
+    bidirectional_iterator<_Iter> ||
+#endif
+    is_convertible_v<_Iter_cat_t<_Iter>, bidirectional_iterator_tag>;
```
Comment (miscco) [inline stl/inc/xutility:866]: Ditto
```suggestion
    _Is_bidi_iter_v<_Iter>;
```

File: stl/inc/algorithm (hunk 6 above)
Comment (miscco) [reply stl/inc/algorithm:4106]: I think this is not correct. This should be `_Is_ranges_fwd_iter_v`

File: stl/inc/functional [hunk 11]
```diff
@@ -2204,7 +2204,9 @@ _CONSTEXPR20 pair<_FwdItHaystack, _FwdItHaystack> _Search_pair_unchecked(
         }
 
         return {_Last1, _Last1};
-    } else if constexpr (_Is_fwd_iter_v<_FwdItHaystack> && _Is_fwd_iter_v<_FwdItPat>) {
```
Comment (miscco) [inline stl/inc/functional:2207]: I am not a fan of those drive by changes. The old style makes it more apparent that everything in the branch is valid and everything else is not.

File: stl/inc/execution (hunk 4 above)
Comment (strega-nil-ms) [reply stl/inc/execution:1579]: Yep.

File: stl/inc/yvals_core.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (strega-nil-ms) [reply stl/inc/yvals_core.h:1351]: It did seem to me like a DR; especially since it's strictly allowing more types in many of these algorithms, I don't feel uncomfortable backporting the changes. If others think it'd be better not to backport, I'm fine with that too, though.

File: stl/inc/algorithm (hunk 6 above)
Comment (strega-nil-ms) [reply stl/inc/algorithm:4106]: It was a broken test before; `_Can_reread_dest` is used on a mutable iterator, and so it should use the Cpp17*Iterator type.

File: stl/inc/algorithm (hunk 7 above)
Comment (strega-nil-ms) [reply stl/inc/algorithm:4411]: Almost certainly.

File: stl/inc/functional (hunk 11 above)
Comment (strega-nil-ms) [reply stl/inc/functional:2207]: I disagree strongly. `_Is_fwd_iter_v<_FwdItHaystack> && _Is_fwd_iter_v<_FwdItPat>` are requirements, and thus are different than the other checks - the other checks just allow us to make the algorithm more efficient, this is necessary for us to do anything at all.

File: tests/std/tests/P2408R5_ranges_iterators_to_classic_algorithms/test.cpp [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        }
+    }
+};
+
+template <class I1, class I2>
+struct binary_algorithms {
+    static void call() {
+        if constexpr (forward_iterator<I1> || _Is_fwd_iter_v<I2>) {
+            helper<I1, I2> h({0, 0, 1, 2, 3, 3, 4, 5});
+            initializer_list exp{0, 1, 2, 3, 4, 5};
```
Comment (frederick-vs-ja) [inline tests/std/tests/P2408R5_ranges_iterators_to_classic_algorithms/test.cpp:92]: This line and similar lines failed due to LLVM-40378... We need to write `initializer_list<int>` to support Clang.

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/functional (hunk 11 above)
Comment (StephanTLavavej) [reply stl/inc/functional:2207]: Either way it's a `static_assert`, so I feel less strongly about this than both of you :joy_cat: That said, I find @strega-nil-ms's rationale for why the checks are different to be persuasive.

This cleanup might have merited a separate PR (as we should generally avoid mixing cleanup work with feature work), but I'm okay with this given how small the change is.

**Comment by StephanTLavavej** [comment]
Thanks, this looks good. I've pushed changes for extremely minor test nitpicks, and a tiny bit of missing coverage in the feature-test macro test (also fixing a pre-existing occurrence).

File: stl/inc/yvals_core.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/yvals_core.h:1352]: The typical reasons to implement a feature downlevel are:

* Invasively affects product code, such that supporting two modes would be a headache, and
* Doesn't disrupt user code that isn't taking advantage of the feature

This has to be balanced against the portability implications of implementing a feature downlevel (i.e. it allows code to be written with MSVC's STL that wouldn't work with libc++ or libstdc++, if they decide differently).

In general, anything that adds user-visible names is something that we guard for new language modes.

The `enable_shared_from_this` overhaul is a good example of something we implemented unconditionally.

Here, I am slightly concerned about the portability implications issue, but it doesn't seem onerous (only individual algorithm calls would be affected), and this is detectable via a feature-test macro *and* is specific to C++20 and above which is quite new (anything that goes back to C++14 would need more caution).

So I am ok with this, but I asked @CaseyCarter in the Discord just to be extra sure.

[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/yvals_core.h:1352]: I agree this is reasonable in C++20 mode since we're not adding new names, and (hopefully) not changing observable behavior for things that already compiled. We should probably poke the super secret "Standard Library Implementors" email list (which I've forgotten the name of) to let folks know we're implementing this in C++20.

File: stl/inc/functional [hunk 13]
```diff
@@ -2204,7 +2204,9 @@ _CONSTEXPR20 pair<_FwdItHaystack, _FwdItHaystack> _Search_pair_unchecked(
         }
```
Comment (CaseyCarter) [inline stl/inc/functional:2204]: Can we never use the random-access iterator path above with iterators that model `random_access_iterator` but whose iterator category is not `random_access_iterator_tag`?

File: stl/inc/vector [hunk 14]
```diff
@@ -1206,12 +1206,7 @@ public:
 
         _Adl_verify_range(_First, _Last);
         const auto _Whereoff = static_cast<size_type>(_Whereptr - _Oldfirst);
-#ifdef __cpp_lib_concepts
-        constexpr bool _Is_fwd = _Is_fwd_iter_v<_Iter> || forward_iterator<_Iter>;
-#else // ^^^ __cpp_lib_concepts ^^^ / vvv !__cpp_lib_concepts vvv
-        constexpr bool _Is_fwd = _Is_fwd_iter_v<_Iter>;
-#endif // ^^^ !__cpp_lib_concepts ^^^
-        if constexpr (_Is_fwd) {
+        if constexpr (_Is_ranges_fwd_iter_v<_Iter>) {
```
Comment (CaseyCarter) [inline stl/inc/vector:1209]: Every time someone touches a container, I think "Great, another merge conflict for the P1206R7 PR." And then I realize that _every_ change is a merge conflict for Stephan's Modules PR, and then I feel really bad for Stephan. (No change required.)

File: stl/inc/xutility [hunk 15]
```diff
@@ -848,9 +848,23 @@ _INLINE_VAR constexpr bool _Is_input_iter_v = is_convertible_v<_Iter_cat_t<_Iter
 template <class _Iter>
 _INLINE_VAR constexpr bool _Is_fwd_iter_v = is_convertible_v<_Iter_cat_t<_Iter>, forward_iterator_tag>;
 
+template <class _Iter>
+_INLINE_VAR constexpr bool _Is_ranges_fwd_iter_v =
+#if defined(__cpp_lib_concepts)
```
Comment (CaseyCarter) [inline stl/inc/xutility:853]: We conventionally `#ifdef MEOW`/`#ifndef MEOW` instead of `#if defined(MEOW)`/`#if !defined(MEOW)`.  (Also on 863.)

File: stl/inc/algorithm [hunk 16]
```diff
@@ -4096,14 +4096,14 @@ namespace ranges {
 } // namespace ranges
 #endif // __cpp_lib_concepts
 
-#ifdef __cpp_lib_concepts
 template <class _InIt, class _OutIt>
-concept _Can_reread_dest = forward_iterator<_OutIt> && same_as<iter_value_t<_InIt>, iter_value_t<_OutIt>>;
-#else // ^^^ defined(__cpp_lib_concepts) / !defined(__cpp_lib_concepts) vvv
-template <class _InIt, class _OutIt>
-_INLINE_VAR constexpr bool _Can_reread_dest = _Is_fwd_iter_v<_OutIt> //
+#if defined(__cpp_lib_concepts)
```
Comment (CaseyCarter) [inline stl/inc/algorithm:4100]: Ditto `#ifdef`.

File: stl/inc/algorithm (hunk 6 above)
Comment (CaseyCarter) [reply stl/inc/algorithm:4106]: Yes, this is fixing damage I did to `std::unique_copy` ~forever ago.

File: stl/inc/algorithm [hunk 17]
```diff
@@ -4727,16 +4729,17 @@ _SampleIt _Sample_selection_unchecked(
...
     if (0 < _Count) {
         auto _UFirst   = _Get_unwrapped(_First);
         auto _ULast    = _Get_unwrapped(_Last);
         using _PopDiff = _Iter_diff_t<_PopIt>;
         _Rng_from_urng<_PopDiff, remove_reference_t<_Urng>> _RngFunc(_Func);
-        if constexpr (_Is_fwd_iter_v<_PopIt>) {
+        if constexpr (_Is_ranges_fwd_iter_v<_PopIt>) {
             // source is forward: use selection sampling (stable)
             using _CT            = common_type_t<_Diff, _PopDiff>;
             const auto _Pop_size = _STD distance(_UFirst, _ULast);
```
Comment (CaseyCarter) [inline stl/inc/algorithm:4745]: Ditto "Are we missing changes to teach `std::distance` about models of `random_access_iterator` that don't have a category that derives from `random_access_iterator_tag` or have no `iterator_category` at all?"

File: stl/inc/execution [hunk 18]
```diff
@@ -1112,7 +1112,7 @@ bool _All_of_family_parallel(_FwdIt _First, const _FwdIt _Last, _Pr _Pred) {
 template <class _ExPo, class _FwdIt, class _Pr, _Enable_if_execution_policy_t<_ExPo> /* = 0 */>
 _NODISCARD bool all_of(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Pr _Pred) noexcept /* terminates */ {
     // test if all elements in [_First, _Last) satisfy _Pred with the indicated execution policy
-    _REQUIRE_PARALLEL_ITERATOR(_FwdIt);
+    _REQUIRE_CONST_PARALLEL_ITERATOR(_FwdIt);
     _Adl_verify_range(_First, _Last);
     auto _UFirst      = _Get_unwrapped(_First);
     const auto _ULast = _Get_unwrapped(_Last);
```
Comment (CaseyCarter) [inline stl/inc/execution:1118]: Ditto have we audited all the machinery used to implement these parallel algorithms to ensure they are prepared for iterators that model a stronger concept than their `iterator_category` and/or have no `iterator_category`?

File: stl/inc/vector (hunk 14 above)
Comment (StephanTLavavej) [reply stl/inc/vector:1209]: :joy_cat: The modules PR-to-be is actually much more resistant to merge conflicts than you might expect, partially because git is great at resolving changes that touch different lines, partially because I generally need to modify only the top-level class/function definition introductions.

(Large-scale code movement *is* a headache.)

File: stl/inc/functional (hunk 13 above)
Comment (strega-nil-ms) [reply stl/inc/functional:2204]: I didn't want to make changes to behavior that weren't specified explicitly. I can do it differently if you think that's best, though.

File: stl/inc/execution [hunk 19]
```diff
@@ -1274,13 +1274,13 @@ using _Parallel_find_results = conditional_t<_Use_atomic_iterator<_FwdIt>, _Para
     _Parallel_choose_min_chunk<_FwdIt>>;
 
 template <class _FwdIt, class _Find_fx>
-struct _Static_partitioned_find2 {
+struct _Static_partitioned_find3 {
```
Comment (StephanTLavavej) [inline stl/inc/execution:1277]: I see that structs are being renamed to:

```
_Static_partitioned_find3
_Static_partitioned_find_end_forward2
_Static_partitioned_find_end_backward3
_Static_partitioned_adjacent_find3
_Static_partitioned_mismatch3
_Static_partitioned_search3
_Static_partitioned_search_n3
_Static_partitioned_is_sorted_until2
_Static_partitioned_is_heap_until2
```

However, I see no motivation for doing so. Am I missing something, or is this a relic of an intermediate edit?

File: stl/inc/xutility [hunk 20]
```diff
@@ -5219,6 +5252,7 @@ _CONSTEXPR20 _FwdIt rotate(_FwdIt _First, _FwdIt _Mid, _FwdIt _Last) {
     // exchange the ranges [_First, _Mid) and [_Mid, _Last)
     // that is, rotates [_First, _Last) left by distance(_First, _Mid) positions
     // returns the iterator pointing at *_First's new home
+    _REQUIRE_CPP17_MUTABLE_ITERATOR(_FwdIt);
```
Comment (StephanTLavavej) [inline stl/inc/xutility:5255]: As we discussed in Discord, the `_REQUIRE_MEOW` checks were intended for parallel algorithms only and should not be applied to classic non-parallel algorithms (as they may break squirrely user code; we have enough trouble when we add optimizations that dispatch on iterator strength).

You removed several occurrences but `rotate(first, mid, last)` here still needs to be fixed. (I am unsure if this is the last one.)

File: stl/inc/execution [hunk 21]
```diff
@@ -3392,12 +3397,13 @@ _FwdIt _Partition_merge(const _FwdIt _False_first, const _FwdIt _True_first, con
     const _Iter_diff_t<_FwdIt> _Count1, const _Iter_diff_t<_FwdIt> _Count2) {
     // Merge partition ranges where [_False_first, _True_first) are falses, [_True_first, _True_last) are trues
     // pre: _Count1 == distance(_False_first, _True_first) && _Count2 == distance(_True_first, _True_last)
+    _REQUIRE_CPP17_MUTABLE_ITERATOR(_FwdIt);
```
Comment (StephanTLavavej) [inline stl/inc/execution:3400]: `_REQUIRE_MEOW` checks are being added to internal helpers, but I believe that our policy is to perform them at the beginning of user-visible parallel algorithms only, so that they are diagnosed as early as possible.

You reverted a couple of these, but `_Partition_merge` is still affected. I am unsure if this is the last occurrence.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/execution (hunk 19 above)
Comment (strega-nil-ms) [reply stl/inc/execution:1277]: (see discord https://discordapp.com/channels/737189251069771789/737189251497721887/1004167710277648384)

Comment (StephanTLavavej) [reply stl/inc/execution:1277]: Resolution: `_Use_atomic_iterator` can return different answers now, so anything that depends on it directly or indirectly (e.g. through `_Parallel_find_results`) should be renamed.

**Review by CaseyCarter — APPROVED** [review]
General comment: I think the wording is strong enough to give us the power to test for `sized_sentinel` and use it to compute distances for sub-`random_access` iterators. The potential for breakage is high and the potential benefit smallish, so I'd be happy to see a followup issue to investigate rather than more changes in this PR.

**Comment by CaseyCarter** [comment]
I note that the proposal is more hardline than what we're implementing; it says "must model `meow_iterator`" where we're enforcing more "must model `meow_iterator` or meet the _`Cpp17MeowIterator`_ requirements". I think I'm fine with this as-is; the requirements are all preconditions rather than constraints, so we're welcome to define behavior for violations.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for implementing this C++23 feature/overhaul! :heart_eyes_cat: :rocket: :white_check_mark:
